### ISSUE TITLE
VMs get stuck at Waiting for SSH when ssh-agent is running and has a ed25519 key

### ISSUE THREAD

**Issue by bowlofeggs** [vagrant-libvirt/vagrant-libvirt#452]
Description of problem:
On Fedora Rawhide I have a very repeatable problem where vagrant-libvirt never thinks the guest VM's ssh server is running and sits forever at "Waiting for SSH to become available...". The SSH server is running in the guest, as I can ssh to it from the host and it answers and allows me to login. Killing ssh-agent works around the issue. This is vagrant-libvirt-0.0.30-2.fc23.noarch.

Reproduction steps:
0) Create an ed25519 ssh key.
1) Start ssh-agent.
2) Add the ed25519 to the running ssh-agent with ssh-add.
3) Use the below example Vagrant file, and vagrant up.

Here is a simple Vagrantfile that demonstrates the issue:

```
# -*- mode: ruby -*-
# vi: set ft=ruby :

VAGRANTFILE_API_VERSION = "2"

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 config.vm.box = "https://download.gluster.org/pub/gluster/purpleidea/vagrant/fedora-22/fedora-22.box"

 config.vm.provider :libvirt do |domain|
 end

 config.vm.provision "shell", inline: "echo \"This won't happen\""
end
```

The downstream bug report is here: https://bugzilla.redhat.com/show_bug.cgi?id=1259559

**Comment by purpleidea** [comment]
As an aside, use: https://dl.fedoraproject.org/pub/alt/purpleidea/vagrant/ instead. Use of the gluster server should be deprecated (301's are in place for now) since apparently all my traffic was throwing off their stats :P

**Comment by purpleidea** [comment]
Also, never saw this issue. Usually it will hang at the IP waiting, so once you see the SSH message, you're usually golden. Can you reproduce with the Centos 7.1 box I've produced?

**Comment by rbarlow** [comment]
James wrote:
> As an aside, use:
> https://dl.fedoraproject.org/pub/alt/purpleidea/vagrant/ instead. Use of
> the gluster server should be deprecated (301's are in place for now)
> since apparently all my traffic was throwing off their stats :P

Thanks for the note! I have updated Pulp to use this new URL in this PR:

https://github.com/pulp/pulp/pull/2023

**Comment by rbarlow** [comment]
James wrote:
> Can you reproduce with the Centos 7.1 box I've produced?

Yes, the same issue occurs with the CentOS 7.1 box.

**Comment by purpleidea** [comment]
On Thu, Sep 3, 2015 at 10:38 AM, Randy Barlow <notifications@github.com>
wrote:

> > Can you reproduce with the Centos 7.1 box I've produced?
>
> Yes, the same issue occurs with the CentOS 7.1 box.
>


Try vlog up and see what it says...

**Comment by rbarlow** [comment]
@Seeohelaien thanks for the tip, I did not know about --debug. I've tried that, and I see this message repeatedly until I kill the process:

```
DEBUG ssh: Checking whether SSH is ready...
 INFO machine: Calling action: read_ssh_info on provider Libvirt (f5c6a9e0-a75c-446c-bd13-ce856448f7b0)
 INFO runner: Preparing hooks for middleware sequence...
 INFO runner: 2 hooks defined.
 INFO runner: Running action: machine_action_read_ssh_info #<Vagrant::Action::Builder:0x0055d8a3a5cfb8>
 INFO warden: Calling IN action: #<Vagrant::Action::Builtin::ConfigValidate:0x0055d8a3a50a88>
 INFO warden: Calling IN action: #<VagrantPlugins::ProviderLibvirt::Action::ConnectLibvirt:0x0055d8a3a50a38>
 INFO warden: Calling IN action: #<VagrantPlugins::ProviderLibvirt::Action::ReadSSHInfo:0x0055d8a3a09390>
 INFO machine: Calling action: read_state on provider Libvirt (f5c6a9e0-a75c-446c-bd13-ce856448f7b0)
 INFO runner: Preparing hooks for middleware sequence...
 INFO runner: 2 hooks defined.
 INFO runner: Running action: machine_action_read_state #<Vagrant::Action::Builder:0x0055d8a3980ea0>
 INFO warden: Calling IN action: #<Vagrant::Action::Builtin::ConfigValidate:0x0055d8a39795d8>
 INFO warden: Calling IN action: #<VagrantPlugins::ProviderLibvirt::Action::ConnectLibvirt:0x0055d8a39795b0>
 INFO warden: Calling IN action: #<VagrantPlugins::ProviderLibvirt::Action::ReadState:0x0055d8a393c480>
 INFO warden: Calling OUT action: #<VagrantPlugins::ProviderLibvirt::Action::ReadState:0x0055d8a393c480>
 INFO warden: Calling OUT action: #<VagrantPlugins::ProviderLibvirt::Action::ConnectLibvirt:0x0055d8a39795b0>
 INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::ConfigValidate:0x0055d8a39795d8>
 INFO warden: Calling OUT action: #<VagrantPlugins::ProviderLibvirt::Action::ReadSSHInfo:0x0055d8a3a09390>
 INFO warden: Calling OUT action: #<VagrantPlugins::ProviderLibvirt::Action::ConnectLibvirt:0x0055d8a3a50a38>
 INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::ConfigValidate:0x0055d8a3a50a88>
DEBUG ssh: Checking key permissions: /home/rbarlow/.vagrant.d/insecure_private_key
 INFO ssh: Attempting SSH connection...
 INFO ssh: Attempting to connect to SSH...
 INFO ssh:   - Host: 192.168.121.51
 INFO ssh:   - Port: 22
 INFO ssh:   - Username: vagrant
 INFO ssh:   - Password? false
 INFO ssh:   - Key Path: ["/home/rbarlow/.vagrant.d/insecure_private_key"]
DEBUG ssh: == Net-SSH connection debug-level log START ==
DEBUG ssh: D, [2015-09-04T14:30:10.617914 #22314] DEBUG -- net.ssh.transport.session[2aec51c0ba94]: establishing connection to 192.168.121.51:22
D, [2015-09-04T14:30:10.618259 #22314] DEBUG -- net.ssh.transport.session[2aec51c0ba94]: connection established
I, [2015-09-04T14:30:10.618342 #22314]  INFO -- net.ssh.transport.server_version[2aec51c0ae78]: negotiating protocol version
D, [2015-09-04T14:30:10.623118 #22314] DEBUG -- net.ssh.transport.server_version[2aec51c0ae78]: remote is `SSH-2.0-OpenSSH_6.8'
D, [2015-09-04T14:30:10.623187 #22314] DEBUG -- net.ssh.transport.server_version[2aec51c0ae78]: local is `SSH-2.0-Ruby/Net::SSH_2.9.1 x86_64-linux'
D, [2015-09-04T14:30:10.626351 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: read 944 bytes
D, [2015-09-04T14:30:10.626433 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: received packet nr 0 type 20 len 940
I, [2015-09-04T14:30:10.626477 #22314]  INFO -- net.ssh.transport.algorithms[2aec51c0a2c0]: got KEXINIT from server
I, [2015-09-04T14:30:10.626549 #22314]  INFO -- net.ssh.transport.algorithms[2aec51c0a2c0]: sending KEXINIT
D, [2015-09-04T14:30:10.626657 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: queueing packet nr 0 type 20 len 2020
D, [2015-09-04T14:30:10.626716 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: sent 2024 bytes
I, [2015-09-04T14:30:10.626733 #22314]  INFO -- net.ssh.transport.algorithms[2aec51c0a2c0]: negotiating algorithms
D, [2015-09-04T14:30:10.626796 #22314] DEBUG -- net.ssh.transport.algorithms[2aec51c0a2c0]: negotiated:
* kex: diffie-hellman-group14-sha1
* host_key: ssh-rsa
* encryption_server: aes128-ctr
* encryption_client: aes128-ctr
* hmac_client: hmac-sha1
* hmac_server: hmac-sha1
* compression_client: none
* compression_server: none
* language_client: 
* language_server: 
D, [2015-09-04T14:30:10.626821 #22314] DEBUG -- net.ssh.transport.algorithms[2aec51c0a2c0]: exchanging keys
D, [2015-09-04T14:30:10.628025 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: queueing packet nr 1 type 30 len 268
D, [2015-09-04T14:30:10.628114 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: sent 272 bytes
D, [2015-09-04T14:30:10.630144 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: read 848 bytes
D, [2015-09-04T14:30:10.630219 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: received packet nr 1 type 31 len 828
D, [2015-09-04T14:30:10.630726 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: queueing packet nr 2 type 21 len 20
D, [2015-09-04T14:30:10.630890 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: sent 24 bytes
D, [2015-09-04T14:30:10.630985 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: received packet nr 2 type 21 len 12
D, [2015-09-04T14:30:10.631290 #22314] DEBUG -- net.ssh.authentication.session[2aec5187b21c]: beginning authentication of `vagrant'
D, [2015-09-04T14:30:10.631391 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: queueing packet nr 3 type 5 len 28
D, [2015-09-04T14:30:10.631417 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: sent 52 bytes
D, [2015-09-04T14:30:10.670685 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: read 52 bytes
D, [2015-09-04T14:30:10.670820 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: received packet nr 3 type 6 len 28
D, [2015-09-04T14:30:10.670874 #22314] DEBUG -- net.ssh.authentication.session[2aec5187b21c]: trying none
D, [2015-09-04T14:30:10.670972 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: queueing packet nr 4 type 50 len 44
D, [2015-09-04T14:30:10.671010 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: sent 68 bytes
D, [2015-09-04T14:30:10.671376 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: read 84 bytes
D, [2015-09-04T14:30:10.671471 #22314] DEBUG -- tcpsocket[2aec51c0b60c]: received packet nr 4 type 51 len 60
D, [2015-09-04T14:30:10.671512 #22314] DEBUG -- net.ssh.authentication.session[2aec5187b21c]: allowed methods: publickey,gssapi-keyex,gssapi-with-mic,password
D, [2015-09-04T14:30:10.671551 #22314] DEBUG -- net.ssh.authentication.methods.none[2aec51872f2c]: none failed
D, [2015-09-04T14:30:10.671583 #22314] DEBUG -- net.ssh.authentication.session[2aec5187b21c]: trying publickey
D, [2015-09-04T14:30:10.671720 #22314] DEBUG -- net.ssh.authentication.agent[2aec51866ab0]: connecting to ssh-agent
D, [2015-09-04T14:30:10.671805 #22314] DEBUG -- net.ssh.authentication.agent[2aec51866ab0]: sending agent request 1 len 44
D, [2015-09-04T14:30:10.671954 #22314] DEBUG -- net.ssh.authentication.agent[2aec51866ab0]: received agent packet 5 len 1
D, [2015-09-04T14:30:10.672031 #22314] DEBUG -- net.ssh.authentication.agent[2aec51866ab0]: sending agent request 11 len 0
D, [2015-09-04T14:30:10.672096 #22314] DEBUG -- net.ssh.authentication.agent[2aec51866ab0]: received agent packet 12 len 93

DEBUG ssh: == Net-SSH connection debug-level log END ==
 INFO ssh: SSH not up: #<Vagrant::Errors::NetSSHException: An error occurred in the underlying SSH library that Vagrant uses.
The error message is shown below. In many cases, errors from this
library are caused by ssh-agent issues. Try disabling your SSH
agent or removing some keys and try again.

If the problem persists, please report a bug to the net-ssh project.

curve name mismatched (`v>����1l߁db}2t�{�-��Q�WB�+�	' with `')>
```

I did as the message suggested and killed by running ssh-agent process that it worked! I am not sure why my ssh-agent would get in the way here though, especially since I am able to ssh to the vagrant box manually and that works fine.

It's mentioning something about a curve name mismatch with some garbage characters. My primary SSH key is a ed25519 key. Does it not like the 25519 curve?

**Comment by purpleidea** [comment]
@rbarlow remember there's also vlog (provided by Oh-My-Vagrant)

```
$ which vlog
vlog ()
{ 
    VAGRANT_LOG=info $VAGRANT "$@" 2> vagrant.log
}
```

**Comment by rbarlow** [comment]
James wrote:
> @rbarlow <https://github.com/rbarlow> remember there's also vlog
> (provided by Oh-My-Vagrant)
> 
> |$ which vlog
> vlog ()
> { 
>     VAGRANT_LOG=info $VAGRANT "$@" 2> vagrant.log
> }
> |

Oh cool, thanks! I hadn't installed omv on this Rawhide box yet since
Vagrant had been giving me trouble. I'll get to that soon though. I've
only been using this computer for two days so I don't have everything
configured the way I like it yet.

**Comment by alff** [comment]
Faced some issue. After some investigations i've found that i don't have 'vagrant-libvirt' network in my network list.
 virsh # net-list 
  Name                 State      Autostart     Persistent
 ----------------------------------------------------------
  default              active     yes           yes
  private              active     yes           yes

I'm using latest one version (vagrant-libvirt (0.0.30)) on centos 7.1. I've tried to re-install this plugin but no luck. My question is: when this net should be created? On which step?

**Comment by pronix** [comment]
I guess https://github.com/pradels/vagrant-libvirt/blob/master/lib/vagrant-libvirt/action/create_networks.rb

**Comment by rbarlow** [comment]
I've updated this bug's title and description to reflect the knowledge that ssh-agent and the ed25519 key are involved.

**Comment by alff** [comment]
My apologies to everyone but it seems that i've found my issue.
Due host setup 'home' folders doesn't belong to users. And when i've tried to spin up VM 'ssh' client tried to pick ssh key which was not owned by vagrant user. And it throw exception and get stuck. After move 'home' folder and changing owner issue was fixed.

**Comment by bowlofeggs** [comment]
Should this issue be reported against Vagrant itself, or it is an issue with vagrant-libvirt?

**Comment by infernix** [comment]
@bowlofeggs i haven't dug into this yet but from the debug log it doesn't look like this is about vagrant; seems to be a net-ssh issue (see https://github.com/net-ssh/net-ssh/issues/191 or https://github.com/mitchellh/vagrant/issues/4367 or https://github.com/StefanScherer/packer-windows/commit/24e04b211d3096df64f215867c389f53e0915d7d) - can you check?

**Comment by bowlofeggs** [comment]
@infernix, thank you for the tip! It seems that my system has net-ssh-1.9.1, where they claim to have fixed the issue in 1.9.2. I [filed a request](https://bugzilla.redhat.com/show_bug.cgi?id=1369491) with the Fedora package maintainer to bump Fedora 24 to 1.9.2. I'm happy enough to assume that will fix my issue for now, so we can close this issue.
